fixes #2343 - override default settings with custom values

tstrachota opened the pull request:
- added check for nil before attribute 'default' is loaded from yaml
- removed logic that tried to set 'value' attr to nil when it was equal to 'default';
  it didn't behave predictably as 'default' can be assigned after 'value'

Commit:
fixes #2343 - override default settings with custom values

- added check for nil before attribute 'default' is loaded from yaml
- removed logic that tried to set 'value' attr to nil when it was equal to 'default';
  it didn't behave predictably as 'default' can be assigned after 'value'

domcleal commented:
@tstrachota any chance of tests for this? :)

tstrachota commented:
I can try to add some:)

tstrachota commented:
Ok, it will take a bit longer than I expected. I discovered more issues with the Settings class that need to be fixed.

Commit:
settings - tests and fixes

- added string settings type
- fixed tests previously commented out due to CI failures
- more tests for core settings functionality
- parsing of string values enhanced and extracted to a separate method
- setting value to nil when it equals default is back

domcleal commented:
@tstrachota getting three failures unfortunately, see the bottom of this page (irritatingly, some gem has broken and it's no longer showing test names..):

http://ci.theforeman.org/job/test_develop_pull_request/1081/database=postgresql,puppet=3.0,ruby=1.8.7/console

tstrachota commented:
Tests fixed

domcleal commented on app/models/setting.rb:
```diff
@@ -89,48 +88,86 @@ def default=(v)
+      elsif val == "false"
+        self.value = false
+      else
+        invalid_value_error "must be boolean"
```
These message should be extracted with _().

tstrachota replied:
good point!

domcleal commented on app/models/setting.rb:
```diff
@@ -89,48 +88,86 @@ def default=(v)
+      raise NotImplementedError, "parsing hash from string is not supported"
+
+    else
+      raise "no settings type set"
```
Use Foreman::Exception here

domcleal commented on app/models/setting.rb:
```diff
@@ -89,48 +88,86 @@ def default=(v)
+
+  def self.create opts
+    if (s = Setting.find_by_name(opts[:name].to_s)).nil?
+      super opts
```
This previously used to call `create!` and so it'd throw an exception in the subclasses (app/models/setting/*.rb) if the setting wouldn't save.  I don't think the return value of this method is checked anywhere, so I don't think we'll catch errors.

tstrachota replied:
Shouldn't we redefine also 'Settings#create!' and use it where we now use 'create' then? I'd rather keep the return values and behaviour as they are defined in ActiveRecord::Base.

domcleal replied:
Agreed, that would be good.  Just bear in mind there are downstream consumers of Setting.create currently, the foreman-katello-engine and one or two of @GregSutcliffe's plugins I think.

GregSutcliffe replied:
Just let me know what I need to update ;)

domcleal commented:
Excellent job @tstrachota, that's a lot more robust, especially the test harness.

Commit:
Setting#create! updates default value for existing setting records

create! is now prefered method for creating settings as it raises exceptions on invalid states

Commit:
Setting takes string as default type when parsing values from user input

tstrachota commented:
I added method 'create!' as I mentioned in comments earlier. @GregSutcliffe can you please update your plugin once we merge this? I made a PR for katello-foreman-engine so katello guys have been notified.

domcleal commented on app/models/setting.rb:
```diff
@@ -80,57 +78,105 @@ def value
+      raise NotImplementedError, "parsing hash from string is not supported"
+
+    else
+      raise Foreman::Exception, "parsing settings type '#{settings_type}' from string is not defined"
```
This can be extracted too, replace with this:

```
raise Foreman::Exception.new(N_("parsing settings type '%s' from string is not defined"), settings_type)
```

domcleal commented on app/models/setting.rb:
```diff
@@ -80,57 +78,105 @@ def value
+      self.value = val.to_s.strip
+
+    when "hash"
+      raise NotImplementedError, "parsing hash from string is not supported"
```
Can you use Foreman::Exception here too please?

domcleal commented:
:+1:, will merge with the above exception fixes, thanks.

domcleal commented:
Merged as 911d3f5 and 8da68b8, thanks @tstrachota!
